### DISCUSSION TITLE
What is a simple way to pass image？

### DISCUSSION THREAD

**Discussion by UserNameIsABC** [elsa-workflows/elsa-core#1344] [Q&A]
What is a concise way to transfer image data between activities in an Elsa workflow?

Currently, I would like to insert a path in the activity "ReadFile", pass the acquired image data to the activity "HTTP Response", and display the image in a web browser.
what should I do?

(I know I keep the Elsa workflow data as a json file, but I want to treat it as object data instead of binary data)

**Comment by sfmskywalker** [comment]
The easiest way is to provide your image data as an activity output. The type could be anything you want, like a `Stream` or a `byte[]`. Just make sure to configure the activity output property from the designer to use a Transient provider for storage so that it does not get serialized with the workflow instance. Same goes for any activities that receive the file data as input properties. Unless of course you access the activity output from workflow expressions only.

I'm currently writing a tutorial which happens to cover file handling within a workflow, so stay tuned. In the meantime, the [accompanying sample project](https://github.com/elsa-workflows/elsa-guides/tree/master/src/ElsaDemos.DocumentManagement) might offer you some ideas already.  It provides three workflows as JSON files that handle uploaded files. One of the workflows loads the file as a stream and sends it as an attachment via email.

**Comment by UserNameIsABC** [comment]
Thank you for your reply.
By selecting the storage activity output as Transient, we confirmed that no serialization occurred.

How do I display image data on a web browser using the HttpResponse function?
I rewrote "Readfile.cs" as shown below, and newly declared "img" so that it can be called in other abilities.

I want to be able to call up the data stored in "img" in a web browser.
(Alternatively, the Bytes data that was originally coded can be displayed in a web browser without serialization.)
![Question](https://user-images.githubusercontent.com/88302322/127944532-9eedac71-8184-4671-bc25-6935079f6e4c.jpg)

I am looking forward to the future functions.
Thank you for an easy-to-use implementation. I always appreciate it.

**Reply by sfmskywalker** [reply]
To display an `<img>` tag in the browser that displayes your image, you basically have two options:

1. You store the file somewhere on e.g. Azure Blob Storage and make sure its publicly accessible. You then use this public URL when writing the `<img>` tag.
2. Alternatively, you can bas64 encode the image bytes and render the `<img>` tag e.g. as follows:

```html
<img src="data:image/png;base64, iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJggg==" alt="Red dot" />
```

With regards to **transferring** your image data from the **Read File** activity as in your example, you definitely don't want to store it in a static property.

Instead, return your image object as an activity output. You can then access this output from the HTTP Response activity. 
For example, if the output property is called `Image` and your **Read File** property is named `ReadFile1`, then you could use the following liquid syntax to access the `Image` property:

```liquid
{{ Activities.ReadFile1.Image }}
```

To render it as an `<img>` tag, you could write the image to a byte array and convert it to base64.
To serialize the image to a byte array, you have (at least) the following options:

1. Instead of returning the `Image` object as output, return its `byte[]` data.
2. Implement a JS function and/or Liquid filter that can convert an `Image` to a `byte[]`.
3. Implement an activity that converts an `Image` to a `byte[]`.

Once you have a `byte[]`, you can use Liquid to base64 it and render the image like this:

```liquid
<img src="data:image/png;base64, {{ Activities.ReadFile1.ImageBytes | base64 }}" />
```

You will probably want to maintain the image type (right now it's hardcoded to `image/png`). There [is a way to](https://stackoverflow.com/a/1397541/690374) get this information from the original `Image`, so the most elegant way IMHO would be to do the following:

1. Implement a Liquid filter called `image_type` that takes as an input an `Image` and returns the type of image (e.g. `png` or `jpeg`).
2. Implement another Liquid filter called `image_bytes` that takes as an input an `Image` and returns a `byte[]`.

And of course bonus points for sending a PR with these nice liquid filters :)

When all is said and done, you would be able to do this:

```liquid
<img src="data:image/{{ Activities.ReadFile1.Image | image_type }};base64, {{ Activities.ReadFile1.Image | image_bytes | base64 }}" />
```
